### PR TITLE
Fix bug where yellow icon is shown when icon tag is empty #5819

### LINKED ISSUE #5819: Default icon is displayed when icon tag is empty
I found a bug when I loaded a KML file containing the following
`<IconStyle>
<Icon></Icon>
    </IconStyle>`

No icon is specified, and if I import that KML into Google Earth, the placemark icon is not shown.  However, when I import the KML into an app using Cesium, a yellow placemark icon is shown.

### PR DESCRIPTION
**Pull request by josh-bernstein**
Fixes #5819

### CODE REVIEW

**Comment by cesium-concierge** [comment]
Welcome to the Cesium community @josh-bernstein!

Can you please send in a [Contributor License Agreement](https://github.com/AnalyticalGraphicsInc/cesium/blob/master/CONTRIBUTING.md#contributor-license-agreement-cla) (CLA) so that we can review and merge this pull request?

I noticed that [CHANGES.md](https://github.com/AnalyticalGraphicsInc/cesium/blob/master/CHANGES.md) has not been updated. If this change updates the public API in any way, fixes a bug, or makes any non-trivial update, please add a bullet point to `CHANGES.md` and comment on this pull request so we know it was updated. For more info, see the [Pull Request Guidelines]( https://github.com/AnalyticalGraphicsInc/cesium/blob/master/CONTRIBUTING.md#pull-request-guidelines).



__I am a bot who helps you make Cesium awesome!__ Thanks again.

**Comment by josh-bernstein** [comment]
My machine's autoformatter made more changes than I meant to

**Comment by josh-bernstein** [comment]
I have updated CHANGES.md.

**Comment by hpinkos** [comment]
Thanks @josh-bernstein! Make sure you send over a CLA so we can review this.  We'll probably also want to add a unit test for this particular case.

@tfili do you want to review?

**Comment by tfili** [comment]
Thanks @josh-bernstein. It looks good, but as Hannah mentioned we would want a unit test. It can probably just load a kml chunk and then verify that there isn't an image in the entity.

**Comment by josh-bernstein** [comment]
@tfili Thanks for the response!  Can you please tell me where the test file is located?

**Comment by biomassives** [comment]
This message was created automatically by mail delivery software.

A message that you sent could not be delivered to one or more of its
recipients. This is a temporary error. The following address(es) deferred:

  acmeideal@gmail.com
    Domain biomassiv.es has exceeded the max emails per hour (106/100 (106%)) allowed.  Message will be reattempted later

------- This is a copy of the message, including all the headers. ------
Received: from github-smtp2-ext5.iad.github.net ([192.30.252.196]:40329 helo=github-smtp2a-ext-cp1-prd.iad.github.net)
	by chi-server32.websitehostserver.net with esmtps (TLSv1.2:ECDHE-RSA-AES256-GCM-SHA384:256)
	(Exim 4.89)
	(envelope-from <noreply@github.com>)
	id 1drnDl-000huE-C2
	for greg@biomassiv.es; Tue, 12 Sep 2017 10:35:29 -0500
Date: Tue, 12 Sep 2017 08:27:53 -0700
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=github.com;
	s=pf2014; t=1505230488;
	bh=e9INILpxZkBGXPEbMcZAhNg+p6SVSbbcwD8PKvYfd+8=;
	h=From:Reply-To:To:Cc:In-Reply-To:References:Subject:List-ID:
	 List-Archive:List-Post:List-Unsubscribe:From;
	b=Yf8vNlKp/LLCFvCYzjGikkKt87n30/fOLfzi0IV6VhZl6eNcGxYeW5j+n2ohWJYpm
	 xXiGQaeU0Q2NZ3uJfv1TmzLDuNstXl/DJBP/AEGdFeqTh8fuF8qrmCK9IsQWW0EUVn
	 7DkTMAbqqMHjiW279TuXsLLX0npnhYgVp4rX/MVU=

**Comment by tfili** [comment]
@josh-bernstein All tests are located in the `Specs` folder at the same location as in the `Source` folder. In this case `Specs/DataSources/KmlDataSourceSpec.js`

**Comment by josh-bernstein** [comment]
@hpinkos @tfili I have added unit test.  I am still waiting on my company to fill out the CLA.

**Comment by mramato** [comment]
@josh-bernstein thanks again for this PR.  Is there any update on the CLA?

**Comment by josh-bernstein** [comment]
@mramato My company says they should be sending it this week!

**Comment by mramato** [comment]
Awesome!  Thanks.

**Comment by pjcozzi** [comment]
We now have a CLA from @josh-bernstein, thanks!

@josh-bernstein can you please add yourself to [CONTRIBUTORS.md](https://github.com/AnalyticalGraphicsInc/cesium/blob/master/CONTRIBUTORS.md) under the corporate CLA section?

File: Source/DataSources/KmlDataSource.js [hunk 1]
```diff
@@ -1111,10 +1111,6 @@ define([
             entity.billboard = billboard;
         }
 
-        if (!defined(billboard.image)) {
```
Comment (tfili) [inline Source/DataSources/KmlDataSource.js]: By doing this, it fixes your case, but breaks the case where there isn't an icon style at all. In that case Google Earth does show a yellow push pin.

Changing the if to something like
```
if (!defined(styleEntity.billboard) && !defined(billboard.image)) {
```
will only set the default image if an icon style wasn't defined, which I believe is the correct behavior in both cases.

File: Specs/DataSources/KmlDataSourceSpec.js [hunk 2]
```diff
@@ -263,6 +263,18 @@ defineSuite([
...
 
+    it('if load contains <icon> tag with no image included, no image is added', function() {
+        var dataSource = new KmlDataSource(options);
+        return loadBlob('Data/KML/simpleNoIcon.kml').then(function(blob) {
+            return dataSource.load(blob);
+        }).then(function(source) {
+            expect(source.entities);
+            expect(source.entities.values.length).toEqual(1);
+            expect(source.entities._entities._array.length).toEqual(1);
+            expect(!source.entities._entities._array[0]._billboard._image);
```
Comment (tfili) [inline Specs/DataSources/KmlDataSourceSpec.js]: You need to expect it to be something. This would probably be `toBeUndefined()`.

Also add a test for the other case where no icon style is specified at all, to verify there is the default push pin.

**Comment by josh-bernstein** [comment]
@pjcozzi @tfili I believe I have made your requested changes.  Please let me know if there are any other issues

**Comment by josh-bernstein** [comment]
@tfili I have added the tests you requested, but they seems to be causing errors with Travis.  They do pass on my local machine.   If you could advise if I need to change anything, I would appreciate it

**Comment by pjcozzi** [comment]
Thanks for the update, @josh-bernstein!

@tfili can you take another look today?  We'll include this in the release if it is ready.

**Comment by josh-bernstein** [comment]
@tfili Just wanted to comment, after fixing merge conflicts Travis tests are now passing.  So you can ignore that part of my previous comment

File: Specs/DataSources/KmlDataSourceSpec.js [hunk 3]
```diff
@@ -263,6 +263,42 @@ defineSuite([
...
+
+    it('if load does not contain icon <style> tag for placemark, default yellow pin does show', function() {
+        var dataSource = new KmlDataSource(options);
+        return loadBlob('Data/KML/simpleNoStyle.kml').then(function(blob) {
+            return dataSource.load(blob);
+        }).then(function(source) {
+            expect(source.entities);
+            expect(source.entities.values.length).toEqual(1);
+            expect(source.entities._entities._array.length).toEqual(1);
+            expect(source.entities._entities._array[0]._billboard._image === dataSource._pinBuilder.fromColor(Color.YELLOW, 64));
```
Comment (tfili) [inline Specs/DataSources/KmlDataSourceSpec.js]: You should use `.toEqual` here as well.

File: Specs/DataSources/KmlDataSourceSpec.js [hunk 4]
```diff
@@ -263,6 +263,42 @@ defineSuite([
...
+
+    it('if load contains empty <IconStyle> tag for placemark, default yellow pin does show', function() {
+        var dataSource = new KmlDataSource(options);
+        return loadBlob('Data/KML/simpleEmptyIconStyle.kml').then(function(blob) {
+            return dataSource.load(blob);
+        }).then(function(source) {
+            expect(source.entities);
+            expect(source.entities.values.length).toEqual(1);
+            expect(source.entities._entities._array.length).toEqual(1);
+            expect(source.entities._entities._array[0]._billboard._image === dataSource._pinBuilder.fromColor(Color.YELLOW, 64));
```
Comment (tfili) [inline Specs/DataSources/KmlDataSourceSpec.js]: Use `.toEqual`

**Comment by tfili** [comment]
Thanks @josh-bernstein. Sorry for the delay. I was on vacation last week.

Just those 2 small changes in the tests and I'll merge it.

**Comment by jburr-nc** [comment]
Thanks @tfili. Josh is on vacation until Monday. Changes.md may also need updating to reflect 1.39.

**Comment by josh-bernstein** [comment]
I'm running into new issues as I try to update the tests.  Closing this PR until I can get them fixed

**Comment by josh-bernstein** [comment]
@tfili I have updated the tests as you requested, and this should be good to go.  Please re-review the code, as I did make an additional change to cover an edge case.

File: Source/DataSources/KmlDataSource.js [hunk 5]
```diff
@@ -701,7 +701,12 @@ define([
         var color = queryColorValue(node, 'color', namespaces.kml);
 
         var iconNode = queryFirstNode(node, 'Icon', namespaces.kml);
-        var icon = getIconHref(iconNode, dataSource, sourceUri, uriResolver, false, query);
+        var icon;
+        if (!defined(iconNode)) {
```
Comment (tfili) [inline Source/DataSources/KmlDataSource.js]: What situation does this address that isn't addressed below?

**Comment by tfili** [comment]
One of the KML tests still fails.

Also, it looks like you removed an existing entry in CHANGES.md when merging it.

File: Source/DataSources/KmlDataSource.js (hunk 5 above)
Comment (josh-bernstein) [reply Source/DataSources/KmlDataSource.js]: The situation where <IconStyle> tags are empty.  For an example, see simpleEmptyIconStyle.kml.  This differs from the situation where <IconStyle> tag includes a blank <Icon> inside, as in simpleNoIcon.kml.

Google Earth shows yellow placemark in the first situation and no placemark in the second.  This code here is necessary to have yellow placemark show in the first situation.

**Comment by josh-bernstein** [comment]
@tfili When I look at my code changes for CHANGES.md, it does not show anything removed.  I did remove my change from the 1.38 version, maybe that's what you're seeing.

Also, my unit tests pass locally.  I see that Travis is failing, but I'm not sure what is causing those errors.

**Comment by tfili** [comment]
@josh-bernstein Are you sure you are running all the tests? It fails for me locally and stepping into the code, I don't know how it would pass anywhere.
In the `Styles: inline styles take precedance over shared styles` test, what is happening is
* The `<IconStyle>` gets processed with the `href` of `http://test.invalid`
* Then the inline style gets processed. This doesn't have a `href`. The styles are merged and this is where you have a problem.
  * In master since `image` is undefined in the inline style, the billboard fallsback to the `IconStyle.image` and is set to `http://test.invalid`
  * In you branch since image is undefined, you set it to the pushpin and that's how it remains, so the test fails because it expects `billboard.image` to be `http://test.invalid` but instead it is a `canvas` element.
* We need to push off figuring out what to use till `processPositionGraphics` once all the potential styles have been merged.
* The cases we have are
  * **CORRECT**
    * Style has no IconStyle - Shows yellow pushpin
      * `entity.billboard` is `undefined`
    * Style has `IconStyle` with `Icon` - Uses the Icon
      * `entity.billboard` contains image
    * Style has `IconStyle` with empty `Icon` - Show nothing
      * `entity.billboard` is defined but with no image
  * **INCORRECT**
    * Style has `IconStyle` with no `Icon` - Shows nothing but should show yellow pushping
      * `entity.billboard` is defined but with no image

My suggestion would be in `processBillboardIcon`, that if you don't have a `Icon` or a `color` then you don't create a billboard at all. Then it would fallback to the yellow pushpin since `entity.billboard` would be undefined. This adds an additional case however, which would be no `Icon` but has a `color`. You would need to account and write a test for this additional case.

---

My mistake about `CHANGES.md`. I guess I was looking at an intermediate change. Looks fine in your branch.

---
You don't need to close the PR when you are working on it. It won't be merged till all the tests pass and comments are addressed anyway. Just comment again once it is ready.

**Comment by josh-bernstein** [comment]
@tfili Thank you for comment.  I believe I have addressed all the concerns you raised.  I have once again made significant code changes to KmlDataSource.js, so please re-review code.

**Comment by josh-bernstein** [comment]
Actually I think I'm still failing this test

> load works with a KMZ file

**Comment by josh-bernstein** [comment]
@tfili I'm not sure why the Travis builds keep failing.  When I look at the errors, none look related to my branch.  Also, when I run the tests locally with master branch checked out, I get errors.  So I do not believe that any of these errors are related to my branch.  Therefore, I believe my branch is good to go.

Please let me know if you have thoughts.

File: Source/DataSources/KmlDataSource.js [hunk 6]
```diff
@@ -729,6 +729,12 @@ define([
 
         var iconNode = queryFirstNode(node, 'Icon', namespaces.kml);
         var icon = getIconHref(iconNode, dataSource, sourceUri, uriResolver, false, query);
+
+        // If icon tags are present but blank, we do not want to show an icon
+        if (defined(iconNode) && !defined(icon)) {
```
Comment (tfili) [inline Source/DataSources/KmlDataSource.js]: Reusing the `icon` variable as a temporary with a different type, even though fine in Javascript, isn't best practice. I would just use a separate variable.

Then you can simplify everything like this
```
var blankIcon = defined(iconNode) && !defined(icon);

...

if (!defined(billboard.image) && !blankIcon) {
              billboard.image = dataSource._pinBuilder.fromColor(Color.YELLOW, 64);
}
```

**Comment by tfili** [comment]
@josh-bernstein Looks good. I've seen that test fail before. I restarted the build. Hopefully it resolves itself. Other than that I just have the one comment on your new code.

File: Source/DataSources/KmlDataSource.js (hunk 6 above)
Comment (josh-bernstein) [reply Source/DataSources/KmlDataSource.js]: blankIcon would be set in the processBillboardIcon method, while the check `if (!defined(billboard.image) && !blankIcon)` would need to happen in processPositionGraphics method.

Therefore, I believe the only ways to incorporate your suggestions would be to make blankIcon a global variable or a property of billboard.  Is either of these ways better than the current way?

**Comment by josh-bernstein** [comment]
@tfili Please see my reply to your comment

**Comment by hpinkos** [comment]
@josh-bernstein I fixed the failing test in master.  If you merge that into your branch, it should pass.  Thanks!

**Comment by josh-bernstein** [comment]
Awesome thanks @hpinkos

**Comment by josh-bernstein** [comment]
Actually @hpinkos it appears that master is already merged into my branch

**Comment by tfili** [comment]
@josh-bernstein My mistake.

Also, it looks like you merged master in 4 hours ago, but the fix wasn't merge until 2 hours ago. Could you merge it in so the tests pass.

**Comment by josh-bernstein** [comment]
@tfili When I do `git pull origin master` it says 'already up-to-date'.
Same if I checkout master, git pull, then go back to my branch and try to merge.

**Comment by hpinkos** [comment]
@josh-bernstein Make sure you're fetching master from our repo: https://help.github.com/articles/syncing-a-fork/

**Comment by josh-bernstein** [comment]
@hpinkos Thank you for the instructions, but I think I'm missing something.

I've tried different combinations of `git fetch upstream`, git fetch AnalyticalGraphicsInc`, `git pull upstream master`, `git pull AnalyticalGraphicsInc:master`, but they all give errors.

Any advice would be appreciated.

Thank you.

**Comment by hpinkos** [comment]
@josh-bernstein make sure you set upstream first
`git remote add upstream git@github.com:AnalyticalGraphicsInc/cesium.git`

**Comment by josh-bernstein** [comment]
@hpinkos @tfili Ok, I merged master.  Let's see if that works.

**Comment by josh-bernstein** [comment]
Oh no, it failed again!

File: Source/DataSources/KmlDataSource.js [hunk 7]
```diff
@@ -1140,6 +1146,8 @@ define([
 
         if (!defined(billboard.image)) {
             billboard.image = dataSource._pinBuilder.fromColor(Color.YELLOW, 64);
+        } else if (billboard.image == false) {
```
Comment (hpinkos) [inline Source/DataSources/KmlDataSource.js]: @josh-bernstein this is why CI failed on eslint.  `==` should be `===`

Comment (josh-bernstein) [reply Source/DataSources/KmlDataSource.js]: @hpinkos This condition is only true in the desired cases if I use ==, not ===.  
I could remove this 'else if' block altogether and it will still function correctly.  I just think it is more intuitive to set the value back from false to undefined

Comment (hpinkos) [reply Source/DataSources/KmlDataSource.js]: Do `} else if (!billboard.image) {` instead

Comment (hpinkos) [reply Source/DataSources/KmlDataSource.js]: Or what is the value of `billboard.image` here?  Is it `'false'` instead of `false`?  
In that case, do `} else if (billboard.image === 'false') {`

Comment (tfili) [reply Source/DataSources/KmlDataSource.js]: You explicitly set `billboard.image` to `false`. `===` should work in this case because the type and value will match.

Comment (josh-bernstein) [reply Source/DataSources/KmlDataSource.js]: @hpinkos @tfili I agree it seems like it should be false.  However, this is the value I get when using my browser's debugger.

billboard

Comment (hpinkos) [reply Source/DataSources/KmlDataSource.js]: Oooh okay.  It's already being converted to an `Property` at this point.
In this case, let's do `} else if (billboard.image instanceof ConstantProperty && !billboard.image.getValue()) {`.

Please also add a comment like you did in your change above so we remember why we did this.

Comment (josh-bernstein) [reply Source/DataSources/KmlDataSource.js]: @hpinkos `billboard.image instanceof ConstantProperty` throws error saying ConstantProperty is undefined.  However, everything seems to work if I do `} else if (billboard.image && !billboard.image.getValue()) {'

Comment (hpinkos) [reply Source/DataSources/KmlDataSource.js]: At the top of the file, require in `ConstantProperty`.  See how we require in `SampledPositionProperty` in this file for an example.  We want to make sure it's a constant property because other property types require an argument passed into `getValue`

Comment (josh-bernstein) [reply Source/DataSources/KmlDataSource.js]: @hpinkos In what other situation would billboard.image be defined but have a value of false?

Comment (hpinkos) [reply Source/DataSources/KmlDataSource.js]: Okay.  I had to double check throughout the code but it looks like if it is defined it will always be a `ConstantProperty`.  So we can safely just do `} else if (!billboard.image.getValue()) {`.  I was just being cautious.

Comment (josh-bernstein) [reply Source/DataSources/KmlDataSource.js]: @hpinkos Ok sounds good.  Thanks for checking it out!

**Comment by josh-bernstein** [comment]
@tfili @hpinkos Looks like it's good now!

**Comment by hpinkos** [comment]
Thanks for fixing this @josh-bernstein!

**Comment by josh-bernstein** [comment]
@hpinkos My pleasure!
